### PR TITLE
ENH: Adds sort_values to dask.DataFrame

### PR DESCRIPTION
**Pull request by gerrymanoim**
- [X] Closes https://github.com/dask/dask/issues/958
- [X] Tests added / passed
- [X] Passes `black dask` / `flake8 dask`. Running black on `core.py` and `shuffle.py` caused large diffs - maybe I have something misconfigured?

As suggested https://github.com/dask/dask/issues/958#issuecomment-573121520 and https://github.com/dask/dask/pull/2367, I've pretty much reused the code in `set_index` and don't mess with the underlying bits. Shared code between `set_index` and `sort_values` is in `_calculate_divisions`. 

Implementation caveats:

- only single column sorting is supported
- only `ascending=True` is supported

Closes https://github.com/dask/dask/issues/958

### CODE REVIEW

**Comment by jsignell** [comment]
This seems like a reasonable solution to me and I feel like we are open to an implementation of `sort_values`

> code duplication

I'd prefer that this much duplication be abstracted away a bit. 

> Running black on core.py and shuffle.py caused large diffs - maybe I have something misconfigured?

We generally use pre-commit to run black. You can read more about how that works here: https://docs.dask.org/en/latest/develop.html#code-formatting

**Comment by gerrymanoim** [comment]
Thanks for the feedback - I'll get this cleaned up.

**Comment by gerrymanoim** [comment]
@jsignell I think this is good to go, let me know if there are changes I should make.

File: dask/dataframe/shuffle.py [hunk 1]
```diff
@@ -437,6 +437,100 @@ def _construct_graph(self):
...
+            mins == sorted(mins)
+            and maxes == sorted(maxes)
+            and all(mx < mn for mx, mn in zip(maxes[:-1], mins[1:]))
+            and npartitions == df.npartitions
+        ):
+            # divisions are in the right place
+            divisions = mins + [maxes[-1]]
+            return df.map_partitions(M.sort_values, value)
+    df = rearrange_by_divisions(df, value, divisions)
+    df.divisions = divisions
```
Comment (jsignell) [inline dask/dataframe/shuffle.py:529]: I don't think you want to set `df.divisions` to `divisions`. The output of `_calculate_division` is the divisions that would be on the `sort_by_col` if it were the index, but in the case of `sort_by_values` the column does not become the index, so the divisions in this method are not equivalent to the divisions on the resultant dataframe.

File: dask/dataframe/tests/test_shuffle.py [hunk 2]
```diff
@@ -1132,3 +1132,18 @@ def test_set_index_nan_partition():
...
+)
+def test_sort_values(npartitions):
+    df = pd.DataFrame({"a": np.random.randint(0, 10, 100)})
+
+    ddf = dd.from_pandas(df, npartitions=npartitions)
+
+    tm.assert_frame_equal(
+        ddf.sort_values("a").compute().reset_index(drop=True),
+        df.sort_values("a").reset_index(drop=True),
+    )
```
Comment (jsignell) [inline dask/dataframe/tests/test_shuffle.py:1149]: It's preferable to use `assert_eq` like the rest of the tests in this file do. That helper function checks properties of the dask dataframe before `compute` is called. By changing this test to use `assert_eq` you'll see the divisions issue that I mentioned above.

**Review by jsignell — COMMENTED** [review]
There are some lingering issues to be resolved, but this PR is coming along nicely!

File: dask/dataframe/tests/test_shuffle.py [hunk 3]
```diff
@@ -1104,9 +1104,9 @@ def test_shuffle_hlg_layer():
 @pytest.mark.parametrize(
     "npartitions",
     [
-        10,  # ShuffleLayer
-        1,  # SimpleShuffleLayer
-    ],
+        10,
+        1,
+    ],  # ShuffleLayer  # SimpleShuffleLayer
```
[the remark this answers is not in the document]
Comment (gerrymanoim) [reply dask/dataframe/tests/test_shuffle.py:1109]: Sorry - something weird happened with autoformatting here.

File: dask/dataframe/core.py [hunk 4]
```diff
@@ -3810,6 +3810,56 @@ def select_dtypes(self, include=None, exclude=None):
...
+        npartitions: int, None, or 'auto'
+            The ideal number of output partitions. If None, use the same as
+            the input. If 'auto' then decide by memory use.
+        divisions: list, optional
+            Known values on which to separate index values of the partitions.
+            See https://docs.dask.org/en/latest/dataframe-design.html#partitions
+            Defaults to computing this with a single pass over the data. Note
+            that if ``sorted=True``, specified divisions are assumed to match
+            the existing partitions in the data. If ``sorted=False``, you should
+            leave divisions empty and call ``repartition`` after ``set_index``.
```
[the remark this answers is not in the document]
Comment (gerrymanoim) [reply dask/dataframe/core.py:3835]: Yeah - you're right here. I was copying the interface from `set_index` and this doesn't make sense.

File: dask/dataframe/tests/test_shuffle.py (hunk 2 above)
Comment (gerrymanoim) [reply dask/dataframe/tests/test_shuffle.py:1149]: Ah! I see - thanks!

**Comment by gerrymanoim** [comment]
This should be in a better state now - thanks for catching those oversights.

**Comment by jakirkham** [comment]
cc @rjzamora (in case this if of interest 🙂)

**Comment by rjzamora** [comment]
> cc @rjzamora (in case this if of interest 🙂)

Seems like the logical way to handle single-column `sort_values` - Thanks for this @gerrymanoim!

Dask-CuDF actually does the same exact thing to support both single- and multi-column `sort_values`.  However, this is only possible with cudf, because the divisions calculation stage is performed with `cudf.DataFrame.quantiles` (which returns a single quantile column for the entire dataframe input, rather than independent quatiles for each column).

**Comment by gerrymanoim** [comment]
> which returns a single quantile column for the entire dataframe input, rather than independent quatiles for each column

Ah that would be a nice extension.

**Comment by jsignell** [comment]
Thanks for the PR @gerrymanoim!

File: dask/dataframe/shuffle.py [hunk 5]
```diff
@@ -437,6 +437,98 @@ def _construct_graph(self):
...
+    )
+
+    if (
+        mins == sorted(mins)
+        and maxes == sorted(maxes)
+        and all(mx < mn for mx, mn in zip(maxes[:-1], mins[1:]))
+        and npartitions == df.npartitions
+    ):
+        # divisions are in the right place
+        divisions = mins + [maxes[-1]]
```
Comment (ryan-williams) [inline dask/dataframe/shuffle.py:524]: just noticing while rebasing #7214: I think this line is dead code.

Not sure if this value was supposed to go anywhere, from skimming the review comments, it seems like it was originally passed to something but is correctly no longer, so this line can just be removed.

File: dask/dataframe/shuffle.py [hunk 6]
```diff
@@ -437,6 +437,98 @@ def _construct_graph(self):
...
+def sort_values(
+    df,
+    value,
+    npartitions=None,
+    ascending=True,
+    upsample=1.0,
+    partition_size=128e6,
+    **kwargs,
+):
+    """ See _Frame.sort_values for docstring """
```
Comment (ryan-williams) [inline dask/dataframe/shuffle.py:492]: I think it ended up as `DataFrame.sort_values`, though moving to `_Frame` does sound like it should "just work" and would be beneficial

Comment (ryan-williams) [reply dask/dataframe/shuffle.py:492]: actually `Series.sort_values()` needs a different API than `DataFrame.sort_values` (former does not take a `by` column name), so someone can add it in another PR; I just updated this comment in #7462

**Pull request merged**